base/sparse/higherorderfns.jl

```diff
@@ -0,0 +1,850 @@
+end
+@inline function _allocres(shape::NTuple{2}, indextype, entrytype, maxnnz)
+    pointers = Vector{indextype}(shape[2] + 1)
+    storedinds = Vector{indextype}(maxnnz)
+    storedvals = Vector{entrytype}(maxnnz)
+    return SparseMatrixCSC(shape..., pointers, storedinds, storedvals)
+end
+# Ambiguity killers, TODO: nix conflicting specializations
+broadcast(::typeof(abs), x::SparseVector) = _noshapecheck_map(abs, x) # base/sparse/sparevectors.jl: 921
+ambiguityfunnel{Tf}(f::Tf, x, y) = _aresameshape(x, y) ? _noshapecheck_map(f, x, y) : _diffshape_broadcast(f, x, y)
+broadcast(::typeof(+), x::SparseVector, y::SparseVector) = ambiguityfunnel(+, x, y) # base/sparse/sparsevectors.jl:1266
+broadcast(::typeof(-), x::SparseVector, y::SparseVector) = ambiguityfunnel(-, x, y) # base/sparse/sparsevectors.jl:1266
+broadcast(::typeof(*), x::SparseVector, y::SparseVector) = ambiguityfunnel(*, x, y) # base/sparse/sparsevectors.jl:1266
+function broadcast!(::typeof(identity), C::SparseMatrixCSC, A::SparseMatrixCSC) # from #17623, loc?
+    _checksameshape(C, A); return copy!(C, A)
+end
+
+
+# (4) _map_zeropres!/_map_notzeropres! specialized for a single sparse vector/matrix
+"Stores only the nonzero entries of `map(f, Array(A))` in `C`."
+function _map_zeropres!{Tf}(f::Tf, C::SparseVecOrMat, A::SparseVecOrMat)
+    spaceC::Int = min(length(storedinds(C)), length(storedvals(C)))
+    Ck = 1
+    @inbounds for j in columns(C)
+        setcolptr!(C, j, Ck)
+        for Ak in colrange(A, j)
+            Cx = f(storedvals(A)[Ak])
+            if Cx != zero(eltype(C))
+                Ck > spaceC && (spaceC = expandstorage!(C, Ck + nnz(A) - (Ak - 1)))
+                storedinds(C)[Ck] = storedinds(A)[Ak]
+                storedvals(C)[Ck] = Cx
+                Ck += 1
+            end
+        end
+    end
+    @inbounds setcolptr!(C, numcols(C) + 1, Ck)
+    trimstorage!(C, Ck - 1)
+    return C
+end
+"""
+Densifies `C`, storing `fillvalue` in place of each unstored entry in `A` and
+`f(A[i])`/`f(A[i,j])` in place of each stored entry `A[i]`/`A[i,j]` in `A`.
+"""
+function _map_notzeropres!{Tf}(f::Tf, fillvalue, C::SparseVecOrMat, A::SparseVecOrMat)
+    # Build dense matrix structure in C, expanding storage if necessary
+    _densestructure!(C)
+    # Populate values
+    fill!(storedvals(C), fillvalue)
+    @inbounds for (j, jo) in zip(columns(C), _densecoloffsets(C))
+        for Ak in colrange(A, j)
+            Cx = f(storedvals(A)[Ak])
+            Cx != fillvalue && (storedvals(C)[jo + storedinds(A)[Ak]] = Cx)
+        end
+    end
+    # NOTE: Combining the fill! above into the loop above to avoid multiple sweeps over /
+    # nonsequential access of storedvals(C) does not appear to improve performance.
+    return C
+end
+# helper functions for these methods and some of those below
+@inline _densecoloffsets(A::SparseVector) = 0
+@inline _densecoloffsets(A::SparseMatrixCSC) = 0:A.m:(A.m*A.n - 1)
```

Comment:
Facepalm-worthy unintentional code obfuscation is why. Your suggestion is *much* better. Fixed on push. Thanks!